File: 3267-find-longest-special-substring-that-occurs-thrice-i/find-longest-special-substring-that-occurs-thrice-i.java
class Solution {
    public int maximumLength(String s) {
        HashMap<String, Integer> memo = new HashMap<String, Integer>();
        for (int i = 0; i < s.length(); i++){
            StringBuilder ss = new StringBuilder();
            for (int j = i; j < s.length(); j++){
                if (ss.length() == 0 || ss.toString().charAt(ss.length() - 1) == s.charAt(j))
                {
                    ss.append(s.charAt(j));
                    memo.put(ss.toString(), 1 + (memo.containsKey(ss.toString()) ? memo.get(ss.toString()) : 0));
                }
                else{
                    break;   
                }
            }
        }
        int result = 0;
        for(Map.Entry<String, Integer> i: memo.entrySet()){
            if (i.getValue() > 2 && result < i.getKey().length())
                result = i.getKey().length();
        }
        return result == 0 ? -1 : result;
    }
}

File: 3267-find-longest-special-substring-that-occurs-thrice-i/find-longest-special-substring-that-occurs-thrice-i.cpp
class Solution {
public:
    int maximumLength(string s) {
        unordered_map<string, int> memo;
        for (int i = 0; i < s.length(); i++){
            stringstream ss;
            for (int j = i; j < s.length(); j++){
                if (ss.str().length() == 0 || ss.str()[ss.str().length() - 1] == s[j])
                {
                    ss << s[j];
                    memo[ss.str()] = 1 + (memo.contains(ss.str()) ? memo[ss.str()] : 0);
                }
                else{
                    break;   
                }
            }
        }
        int result = 0;
        for(const auto& i: memo){
            if (i.second > 2 && result < i.first.length())
                result = i.first.length();
        }
        return result == 0 ? -1 : result;
    }
};